ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java

```diff
@@ -189,8 +190,25 @@ public IRubyObject index(ThreadContext context, IRubyObject key) {
      */
     @JRubyMethod(name = "==")
     public IRubyObject eq(ThreadContext context, IRubyObject _other) {
-        if (_other instanceof RubyHash)
-            return toHash(context).op_equal(context, _other);
+        if (_other instanceof RubyHash){
+            // quick check to see if they are equivalent
+            IRubyObject matched = toHash(context).op_equal(context, _other);
+            if (matched == context.runtime.getTrue()) {
+                return context.runtime.getTrue();
+            }
+
+            RubyHash other = (RubyHash)_other;
```

Comment:
hey @cfallin 

I did not first write a failing test case and then fix it.  The failure was already there @ `ruby/tests/basic.rb#ln616`

that assert fails without the change to `==`.  If you take out this code, you'll see the test case fail; it passes with this deeper logic.  

the root issue is the modifications I made to `#toHash`, which will convert complex child objects (Map/Message) into hashes rather than return the complex object.

I think this 'new' behavior is the expected behavior, but of course that is open to debate :)